File: core-scala/src/main/java/io/asterisque/core/wire/netty/NettyBridge.java
package io.asterisque.core.wire.netty;

import io.asterisque.Asterisque;
import io.asterisque.core.codec.MessageCodec;
import io.asterisque.core.wire.Bridge;
import io.asterisque.core.wire.Server;
import io.asterisque.core.wire.UnsupportedProtocolException;
import io.asterisque.core.wire.Wire;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.handler.ssl.ClientAuth;
import io.netty.handler.ssl.JdkSslContext;
import io.netty.handler.ssl.SslContext;
import io.netty.handler.ssl.SslContextBuilder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLException;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.URI;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;

public class NettyBridge implements Bridge {
  private static final Logger logger = LoggerFactory.getLogger(NettyBridge.class);

  /**
   * この Bridge から生成される Wire で共通して使用されるイベントループです。
   */
  @Nonnull
  private final NioEventLoopGroup worker;

  /**
   * この Bridge に終了命令が出されていることを示すブール値です。
   */
  private final AtomicBoolean closing = new AtomicBoolean(false);

  public NettyBridge() {
    this.worker = new NioEventLoopGroup();
  }

  /**
   * @param uri         接続先の URI
   * @param subprotocol asterisque 上で実装しているサブプロトコル
   * @param sslContext  Secure ソケットを示す URI スキーマが指定された場合 (例えば {@code wss://}) に使用する SSL コンテキスト
   * @return Wire の Future
   */
  @Override
  @Nonnull
  public CompletableFuture<Wire> newWire(@Nonnull URI uri, @Nonnull String subprotocol,
                                         int inboundQueueSize, int outboundQueueSize,
                                         @Nullable SSLContext sslContext) {
    return getScheme(uri).thenCompose(scheme -> {
      if (scheme.equals("ws") || scheme.equals("wss")) {
        WebSocketWire wire = new WebSocketWire(
            MessageCodec.MessagePackCodec, false, inboundQueueSize, outboundQueueSize);
        SslContext ssl = null;
        if (scheme.equals("wss")) {
          ssl = Optional.ofNullable(sslContext)
              .map(ctx -> (SslContext) new JdkSslContext(ctx, true, ClientAuth.NONE))
              .orElse(getDefaultClientSSL());
        }
        WebSocket.Client driver = new WebSocket.Client(worker(), subprotocol, wire.servant, ssl);
        driver.connect(uri);
        return wire.future;
      }

      return Asterisque.Future.fail(new UnsupportedProtocolException("unsupported uri scheme: " + uri));
    });
  }

  /**
   * @param uri         接続先の URI
   * @param subprotocol asterisque 上で実装しているサブプロトコル
   * @param sslContext  Secure ソケットを示す URI スキーマが指定された場合 (例えば {@code wss://}) に使用する SSL コンテキスト
   * @param onAccept    サーバが接続を受け付けたときのコールバック
   * @return Server の Future
   */
  @Override
  @Nonnull
  public CompletableFuture<Server> newServer(@Nonnull URI uri, @Nonnull String subprotocol,
                                             int inboundQueueSize, int outboundQueueSize,
                                             @Nullable SSLContext sslContext,
                                             @Nonnull Consumer<CompletableFuture<Wire>> onAccept) {
    return getScheme(uri).thenCompose(scheme -> {
      if (scheme.equals("ws") || scheme.equals("wss")) {
        WebSocketServer server = new WebSocketServer();
        SslContext ssl = null;
        if (scheme.equals("wss")) {
          ssl = Optional.ofNullable(sslContext)
              .map(ctx -> (SslContext) new JdkSslContext(ctx, false, ClientAuth.NONE))
              .orElseThrow(() -> new NullPointerException("ssl context isn't specified"));
        }
        String path = Optional.ofNullable(uri.getPath())
            .filter(p -> p.length() != 0)
            .orElse("/");
        WebSocket.Server driver = new WebSocket.Server(worker(), subprotocol, path, server, ssl);

        return driver.bind(uriToSocketAddress(uri), channel -> {
          WebSocketWire wire = new WebSocketWire(
              MessageCodec.MessagePackCodec, true, inboundQueueSize, outboundQueueSize);
          onAccept.accept(wire.future);
          return wire.servant;
        }).thenApply(ch -> server);
      }

      return Asterisque.Future.fail(new UnsupportedProtocolException("unsupported uri scheme: " + uri));
    });
  }

  @Override
  public void close() {
    logger.trace("close()");
    if (closing.compareAndSet(false, true)) {
      worker.shutdownGracefully();
    }
  }

  @Nonnull
  private CompletableFuture<String> getScheme(@Nonnull URI uri) {
    if (closing.get()) {
      return Asterisque.Future.fail(new IOException("bridge has been wsClosed"));
    }
    String scheme = uri.getScheme();
    if (scheme == null) {
      return Asterisque.Future.fail(new IllegalArgumentException("uri scheme was not specified: " + uri));
    }
    return CompletableFuture.completedFuture(scheme.toLowerCase());
  }

  @Nonnull
  private SslContext getDefaultClientSSL() {
    try {
      return SslContextBuilder.forClient().build();
    } catch (SSLException ex) {
      throw new IllegalStateException("default client ssl/tls is not available", ex);
    }
  }

  @Nonnull
  private NioEventLoopGroup worker() {
    if (closing.get()) {
      throw new IllegalStateException("netty bridge has been wsClosed");
    }
    return worker;
  }

  @Nonnull
  private InetSocketAddress uriToSocketAddress(@Nonnull URI uri) {
    if (uri.getHost() == null || uri.getHost().length() == 0) {
      return new InetSocketAddress(uri.getPort());
    }
    return new InetSocketAddress(uri.getHost(), uri.getPort());
  }

}

File: core-scala/src/main/scala/io/asterisque/wire/gateway/netty/NettyBridge.scala
package io.asterisque.wire.gateway.netty

import java.io.IOException
import java.net.URI
import java.util.Objects
import java.util.concurrent.atomic.AtomicBoolean

import io.asterisque.wire.gateway.Bridge.Config
import io.asterisque.wire.gateway.{Bridge, Server, UnsupportedProtocolException, Wire}
import io.netty.channel.nio.NioEventLoopGroup
import io.netty.handler.ssl.{ClientAuth, JdkSslContext, SslContext, SslContextBuilder}
import javax.annotation.Nonnull
import javax.net.ssl.{SSLEngine, SSLException}
import org.slf4j.LoggerFactory

import scala.collection.JavaConverters._
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

class NettyBridge() extends Bridge with AutoCloseable {

  def supportedURISchemes:Set[String] = Set("ws", "wss")

  /**
    * この Bridge から生成される Wire で共通して使用されるイベントループです。
    */
  private[this] val _worker = new NioEventLoopGroup

  /**
    * この Bridge に終了命令が出されていることを示すブール値です。
    */
  private[this] val closing = new AtomicBoolean(false)

  /**
    * @param uri    接続先の URI
    * @param config 接続設定
    * @return Wire の Future
    */
  @Nonnull
  override def newWire(@Nonnull uri:URI, @Nonnull config:Config):Future[Wire] = {
    Objects.requireNonNull(uri)
    Objects.requireNonNull(config)
    getScheme(uri) match {
      case Right(secure) =>
        val wire = new WebSocketWire("C:" + uri, primary = false, config.inboundQueueSize, config.outboundQueueSize)

        SslContextBuilder
          .forClient()
          .build()

        val ssl:SslContext = if(secure) {
          Option(config.sslContext)
            /*
            SSLContext sslContext,
                         boolean isClient,
                         Iterable<String> ciphers,
                         CipherSuiteFilter cipherFilter,
                         ApplicationProtocolConfig apn,
                         ClientAuth clientAuth,
                         String[] protocols,
                         boolean startTls
             */
            .map(ctx => new JdkSslContext(ctx, true, ClientAuth.NONE).asInstanceOf[SslContext])
            .getOrElse {
              NettyBridge.logger.debug(s"ssl context ism't specified for uri $uri, use default client ssl")
              getDefaultClientSSL
            }
        } else null
        val driver = new WebSocket.Client(worker, config.subprotocol, ssl)
        driver.connect(uri, _ => wire.servant)
        wire.future
      case Left(ex) => Future.failed(ex)
    }
  }

  /**
    * @param uri      接続先の URI
    * @param config   接続設定
    * @param onAccept サーバが接続を受け付けたときのコールバック
    * @return Server の Future
    */
  @Nonnull
  override def newServer(@Nonnull uri:URI, @Nonnull config:Config, @Nonnull onAccept:Future[Wire] => Unit):Future[Server] = {
    Objects.requireNonNull(uri)
    Objects.requireNonNull(config)
    Objects.requireNonNull(onAccept)
    getScheme(uri) match {
      case Right(secure) =>
        val server = new WebSocketServer(uri)
        val ssl = if(secure) {
          Option(config.sslContext)
            .map(ctx => new JdkSslContext(ctx, false, ClientAuth.NONE).asInstanceOf[SslContext])
            .getOrElse {
              throw new NullPointerException(s"ssl context isn't specified for uri: $uri")
            }
        } else null
        val driver = new WebSocket.Server(worker, config.subprotocol, server, ssl)
        driver.bind(uri, { _ =>
          val wire = new WebSocketWire("S:" + uri, true, config.inboundQueueSize, config.outboundQueueSize)
          onAccept(wire.future)
          wire.servant
        }).map(_ => server)
      case Left(ex) => Future.failed(ex)
    }
  }

  def close():Unit = if(closing.compareAndSet(false, true)) {
    NettyBridge.logger.trace("close()")
    worker.shutdownGracefully()
  }

  @Nonnull
  private def getScheme(@Nonnull uri:URI):Either[Throwable, Boolean] = if(closing.get) {
    Left(new IOException("bridge has been closed"))
  } else {
    Option(uri.getScheme).map(_.toLowerCase) match {
      case Some("ws") => Right(false)
      case Some("wss") => Right(true)
      case None => Left(new IllegalArgumentException(s"uri scheme was not specified: $uri"))
      case _ => Left(new UnsupportedProtocolException(s"unsupported uri scheme: $uri"))
    }
  }

  @Nonnull
  private[this] def getDefaultClientSSL:SslContext = try {
    SslContextBuilder.forClient.build
  } catch {
    case ex:SSLException =>
      throw new IllegalStateException("default client ssl/tls is not available", ex)
  }

  @Nonnull
  private[this] def worker:NioEventLoopGroup = {
    if(closing.get) {
      throw new IllegalStateException("netty bridge has been wsClosed")
    }
    _worker
  }
}

object NettyBridge {
  private[NettyBridge] val logger = LoggerFactory.getLogger(classOf[NettyBridge])

  private[netty] def dump(context:SslContext):Unit = if(context != null && logger.isDebugEnabled) {
    val prefix = if(context.isServer) "SERVER" else if(context.isClient) "CLIENT" else "UNKNOWN"
    logger.debug(f"$prefix: Cipher Suites: ${context.cipherSuites().asScala.sorted.mkString(", ")}")
    logger.debug(f"$prefix: Session Cache Size: ${context.sessionCacheSize()}%,d bytes")
    logger.debug(f"$prefix: Session Timeout: ${context.sessionTimeout()}%,d seconds")
  }

  private[netty] def dump(engine:SSLEngine):Unit = if(engine != null && logger.isDebugEnabled) {
  }
}
